Fix histogram bug with non equal bins

zhujun98 opened the pull request:
# Checklist

- [x ] The title and commit message(s) are descriptive.
- [x ] Small commits made to fix your PR have been squashed to avoid history pollution.
- [x ] Tests have been added for new features or bug fixes.
- [ ] API of new functions and classes are documented.

# Description

<!---
Give any relevant description here. 
If your PR fixes an issue, please include "Fixes #ISSUE" (substituting the relevant issue ID).
-->

Fix #2097

The output of the newly added test before the fix is shown below. The counts in the first and the last bins are wrong.
```
  count
    Which is: { 237, 13, 24, 18, 24, 11, 19, 21, 23, 21, 18, 18, 16, 16, 11, 24, 23, 20, 19, 21, 24, 20, 27, 17, 14, 19, 14, 18, 21, 229 }
  xt::cast<double>(py_count)
    Which is: { 14, 13, 24, 18, 24, 11, 19, 21, 23, 21, 18, 18, 16, 16, 11, 24, 23, 20, 19, 21, 24, 20, 27, 17, 14, 19, 14, 18, 21, 34 }
```

zhujun98 commented:
I wonder whether we can replace the following bound check
```
        // - bounds
        XTENSOR_ASSERT(left <= xt::amin(data)[0]);
        XTENSOR_ASSERT(right >= xt::amax(data)[0]);
```
to 
```
XTENSOR_ASSERT(left <= right)` 
```

It is pretty common to have `left > xt::amin(data)[0]` and `right < xt::amax(data)[0]`.

JohanMabille commented:
Looks like the asserts on bounds (before your change) are more restrictive than what numpy allows, right? It is fine to pass `np.histogram` a data set with values out of the bins, so it should be doable with xtensor too.

tdegeus commented:
The changes make sense to me, thanks a lot! I was just wondering about the test: weren't they partly auto-generated? In that case, the class might break that feature...

zhujun98 commented:
> Looks like the asserts on bounds (before your change) are more restrictive than what numpy allows, right? It is fine to pass `np.histogram` a data set with values out of the bins, so it should be doable with xtensor too.

You are right! It seems to me that the very original implementations of the `histogram` series all assume that the data only contain values within the bins. And it is the source of all the bugs.

zhujun98 commented:
> The changes make sense to me, thanks a lot! I was just wondering about the test: weren't they partly auto-generated? In that case, the class might break that feature...

Good point! But the hand-crafted test did run in the pipeline

https://dev.azure.com/xtensor-stack/xtensor-stack/_build/results?buildId=1994&view=logs&j=e0c44420-fc0a-5f50-ddde-56ed20bc5325&t=fa6c16fe-8d51-5374-9c31-89dfe1ab0f04&l=2484

zhujun98 commented:
@tdegeus Should we fix the `CMakeLists.txt` file to make all the Python related test automatically generated?

tdegeus commented:
@zhujun98 Honestly I forgot how it worked ;) also because I did not design it. Maybe @wolfv or @JohanMabille could comment?

JohanMabille commented:
If I remember correctly, the tests are generated once and then committed. We should change the files generating the tests only, not the generated tests themselves (all the `test_extended_xxx.cpp` files). Adding a warning in these files may help remembering this ;)

> Should we fix the CMakeLists.txt file to make all the Python related test automatically generated?

Yes I think it could be nice. Adding a comment stating that this target (`preprocess_cppy`) should not be run on the CI, but only when whe modify the test source files may be a good thing too.

zhujun98 commented:
I modified `preprocess.py` to generate a warning line.

I would suggest to make another commit to take a close look at other files because they could be modified manually in the past, e.g. the license, which I have fixed all the `cppy` source files.

Regarding the new histogram test with range, it is now generated by the `cppy` file.

tdegeus commented:
> I modified `preprocess.py` to generate a warning line.
> 

Awesome!

> I would suggest to make another PR to take a close look at other `cppy` files because they could be modified manually in the past, e.g. the license, which I have fixed all the `cppy` source files.

That would be a great idea! If you feel up to it, it would be great. Otherwise maybe you can submit an issue such that we at least not forget.

> 
> Regarding the new histogram test with range, it is now generated by the `cppy` file.

Great!

-------

Thanks a lot for the PR!!!

zhujun98 commented:
Thanks a lot for the review! 

> That would be a great idea! If you feel up to it, it would be great. Otherwise maybe you can submit an issue such that we at least not forget.

It is ongoing in #2313.
